docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,84 @@
+
+There are three possible scenarios:
+
+1. The list match: all good
+2. All the blocks known by the querier are within the list of blocks returned by store-gateway, but the store-gateway also included blocks unknown to the querier: all good (it means the store-gateways have discovered and loaded new blocks before the querier discovered them)
+3. Some blocks known by the querier are **not** within the list of blocks returned by store-gateway: potential consistency issue
+
+We want to protect from a partial results response which may occur in the case #3. However, there are some legit cases which, if not handled, would lead to frequent false positives. Given the querier and store-gateway instances independently scan the bucket at a regular interval (to find new blocks or deleted blocks), we may be in one of the following cases:
+
+a. The querier has discovered new blocks before the store-gateway successfully discovered and loaded them
+b. The store-gateway has offloaded blocks "marked for deletion" before the querier
+
+To protect from case (a), we can exclude the blocks which have been uploaded in the last `X` time from the consistency check (same technique already used in other Thanos components). This `X` delay time is used to give the store-gateway enough time to discover and load new blocks, before the querier consider them for the consistency check. This value `X` should be greater than the `-experimental.tsdb.bucket-store.consistency-delay`, because we do expect the querier to consider a block for consistency check once it's reasonably safe to assume that its store-gateway already loaded it.
+
+To protect from case (b) we need to understand how blocks are offloaded. The `BucketStore` (running within the store-gateway) offloads a block as soon as it's not returned by the `MetaFetcher`. This means we can configure the `MetaFetcher` with a [`IgnoreDeletionMarkFilter`](https://github.com/thanos-io/thanos/blob/4bd19b16a752e9ceb1836c21d4156bdeb517fe50/pkg/block/fetcher.go#L648) with a delay of `X` (could be the same value used for case (a)) and in the querier exclude the blocks which have been marked for deletion more than `X` time ago from the consistency check.
+
+## Trade-offs
+
+The proposed solution comes with the following trade-offs:
+
+- A querier is not ready until it has completed an initial full scan of the bucket, downloading the `meta.json` file of every block
+- A store-gateway is not ready until it has completed an initial full scan of the bucket, downloading the `meta.json` and index header of each block matching its shard
+- If a querier hits 2+ store-gateways it may receive duplicated series if the 2+ store-gateways share some blocks due to the replication factor
```

Comment:
That's fine - During the work on chunk iterator I touch merge series Set implementation and we plan lot's of optimizations ther for **exactly** the same series duplicates etc. :hugs:

Reply:
We're also already doing this de-duplication in the Cortex querier (at query time) because we need to de-duplicate series merged between ingesters and storage.